sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -315,4 +316,142 @@ def test_pinjoint_pi():
     _, _, P, C = _generate_body()
     J = PinJoint('J', P, C, parent_axis=P.frame.x+P.frame.y+P.frame.z,
                 child_axis=-C.frame.x-C.frame.y-C.frame.z)
-    assert J._generate_vector() == P.frame.x
+    assert J._generate_vector() == P.frame.y - P.frame.z
+
+
+def test_slidingjoint():
+    _, _, P, C = _generate_body()
+    x, v = dynamicsymbols('x_S, v_S')
+    S = PrismaticJoint('S', P, C)
+    assert S.name == 'S'
+    assert S.parent == P
+    assert S.child == C
+    assert S.coordinates == [x]
+    assert S.speeds == [v]
+    assert S.kdes == [v - x.diff(t)]
+    assert S.parent_axis == P.frame.x
+    assert S.child_axis == C.frame.x
+    assert S.child_point.pos_from(C.masscenter) == Vector(0)
+    assert S.parent_point.pos_from(P.masscenter) == Vector(0)
+    assert S.parent_point.pos_from(S.child_point) == - x * P.frame.x
+    assert P.masscenter.pos_from(C.masscenter) == - x * P.frame.x
+    assert C.masscenter.vel(P.frame) == v * P.frame.x
+    assert P.ang_vel_in(C) == 0
+    assert C.ang_vel_in(P) == 0
+    assert S.__str__() == 'PrismaticJoint: S  parent: P  child: C'
+
+    N, A, P, C = _generate_body()
+    l, m = symbols('l m')
+    S = PrismaticJoint('S', P, C, parent_joint_pos= l * P.frame.x,
+                    child_joint_pos= m * C.frame.y, parent_axis = P.frame.z)
+
+    assert S.parent_axis == P.frame.z
+    assert S.child_point.pos_from(C.masscenter) == m * C.frame.y
+    assert S.parent_point.pos_from(P.masscenter) == l * P.frame.x
+    assert S.parent_point.pos_from(S.child_point) == - x * P.frame.z
+    assert P.masscenter.pos_from(C.masscenter) == - l*N.x - x*N.z + m*A.y
+    assert C.masscenter.vel(P.frame) == v * P.frame.z
+    assert C.ang_vel_in(P) == 0
+    assert P.ang_vel_in(C) == 0
+
+    _, _, P, C = _generate_body()
+    S = PrismaticJoint('S', P, C, parent_joint_pos= l * P.frame.z,
+                    child_joint_pos= m * C.frame.x, parent_axis = P.frame.z)
+    assert S.parent_axis == P.frame.z
+    assert S.child_point.pos_from(C.masscenter) == m * C.frame.x
+    assert S.parent_point.pos_from(P.masscenter) == l * P.frame.z
+    assert S.parent_point.pos_from(S.child_point) == - x * P.frame.z
+    assert P.masscenter.pos_from(C.masscenter) == (-l - x)*P.frame.z + m*C.frame.x
+    assert C.masscenter.vel(P.frame) == v * P.frame.z
+    assert C.ang_vel_in(P) == 0
+    assert P.ang_vel_in(C) == 0
+
+
+def test_slidingjoint_arbitrary_axis():
+    x, v = dynamicsymbols('x_S, v_S')
+
+    N, A, P, C = _generate_body()
+    PrismaticJoint('S', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == -pi
```

Comment:
Why would this return -pi? The function should return an angle that satisfies: 0 <= angle <= pi

Reply:
Because it was considering the above case to - `-(A.x.angle_between(N.x))`

Reply:
Fixed!

Reply:
Please see me comment in the new code.